Shoko.Server/Utilities/TagFilter.cs

```diff
@@ -1,597 +1,607 @@
+            _tagReplacementDictionary.Add("alternative past", existing);
+        }
+
+        /// <summary>
+        /// T needs to have a T(string name) constructor
+        /// </summary>
+        /// <param name="flags"></param>
+        /// <param name="tags"></param>
+        /// <returns></returns>
+        public List<T> ProcessTags(TagFilter.Filter flags, List<T> tags)
+        {
+            if (tags.Count == 0) return tags;
+
+            // We assume the only tag is not "new", so we skip the add/remove logic.
+            if (tags.Count == 1)
+            {
+                if (TagFilter.IsTagBlackListed(_nameSelector(tags[0]), flags)) tags.Clear();
+                return tags;
+            }
+
+            var toAdd = new HashSet<T>(1);
+            var toRemove = new HashSet<T>((int)Math.Ceiling(tags.Count / 2D));
+            var nameToTagDictionary = tags.ToDictionary(_nameSelector, t => t);
+            tags.AsParallel().ForAll(tag => MarkTagsForRemoval(tag, flags, toRemove, toAdd));
+
+            foreach (var tag in toRemove)
+            {
+                // Skip if we want to both remove and add the tag
+                if (toAdd.Contains(tag))
+                    continue;
+
+                tags.Remove(tag);
+            }
+
+            // Add the "original work" tag if no source tags are present and we either want to only include the source tags or want to not exclude the source tags.
+            if (flags.HasFlag(TagFilter.Filter.Source) == flags.HasFlag(TagFilter.Filter.Invert) && !nameToTagDictionary.Any(tag => TagFilter.TagBlackListSource.Contains(tag.Key)))
+            {
+                if (_tagRenameDictionary.TryGetValue("new", out var existing))
+                    toAdd.Add(existing);
+            }
+
+            foreach (var tag in toAdd)
+            {
+                // Skip if we want to both remove and add the tag
+                if (toRemove.Contains(tag))
+                    continue;
+
+                tags.Add(tag);
+            }
+
+            return tags;
+        }
+
+        private void MarkTagsForRemoval(T sourceTag, TagFilter.Filter flags, HashSet<T> toRemove, HashSet<T> toAdd)
+        {
+            var sourceName = _nameSelector(sourceTag);
+            string tag = sourceName.Trim().ToLowerInvariant();
+            if (TagFilter.IsTagBlackListed(tag, flags))
+            {
+                lock (toRemove)
+                {
+                    toRemove.Add(sourceTag);
+                }
+            }
+            else
+            {
+                if (_tagRenameDictionary.TryGetValue(tag, out var newTag))
+                {
+                    lock (toRemove)
+                    {
+                        toRemove.Add(sourceTag);
+                    }
+
+                    lock (toAdd)
+                    {
+                        toAdd.Add(newTag);
+                    }
+                }
+
+                if (_tagReplacementDictionary.TryGetValue(tag, out newTag))

```

Comment:
Why is the dictionary named "\_tag**Replacement**Dictionary" if it is a dictionary of tags to remove if some other tags are present?

Reply:
Because I couldn't think of a good name. Meh